### PR TITLE
[Feat][Plugin] Enable MTP for vLLM Plugin

### PR DESCRIPTION
**Pull request by whx-sjtu**
## Motivation
This PR enables MTP feature for running DeepSeekV3 and GLM5 with vLLM+atom.

## Technical Details

1. `atom_config` related bugfix.
2. Fix wrong `full_cls_name` of different MLA sparse attention backends.
3. Register model architecture and model class for DeepSeek V3 and GLM5 MTP.
4. Add `index_buffer` for `DeepseekMTP`.
5. Adapt full graph of main model with mtp enabled.

## Test Plan

Comming soon.

## Test Result
1. zai-org/GLM-5.1-FP8

Accuracy test commands:
```
lm_eval --model local-completions \
        --model_args model=/home/models/GLM-5.1-FP8,base_url=http://localhost:8000/v1/completions,num_concurrent=64,max_retries=3 \
        --tasks gsm8k \
        --num_fewshot 20
```

Accuracy test result with mtp=3:
TP4
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|    20|exact_match|↑  |0.9454|±  |0.0063|
|     |       |strict-match    |    20|exact_match|↑  |0.9462|±  |0.0062|
```
TP8
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|    20|exact_match|↑  |0.9371|±  |0.0067|
|     |       |strict-match    |    20|exact_match|↑  |0.9386|±  |0.0066|
```
Acceptance rate for mtp=3:
Per-position: **0.934, 0.672, 0.193**, Avg: **60.0%**

2. deepseek-ai/DeepSeek-R1-0528

Accuracy test commands:
```
lm_eval --model local-completions \
        --model_args model=/home/models/DeepSeek-R1-0528,base_url=http://localhost:8000/v1/completions,num_concurrent=16,max_retries=3,tokenized_requests=False \
        --tasks gsm8k \
        --num_fewshot 3
```

Accuracy test result with mtp=3:
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     3|exact_match|↑  |0.9492|±  |0.0060|
|     |       |strict-match    |     3|exact_match|↑  |0.9469|±  |0.0062|
```

## Submission Checklist

- [x] This PR is based on PR #399 and should be merged after that.

### CODE REVIEW

**Comment by wuhuikx** [comment]
Could you please help attach the accuracy test results on gms8k? Do we support MTP=1 or MTP=1/2/3? How about the acceptance ratio?

**Comment by wuhuikx** [comment]
I will turn this PR to draft and go through CI after the code review is done.

**Comment by whx-sjtu** [comment]
> Could you please help attach the accuracy test results on gms8k? Do we support MTP=1 or MTP=1/2/3? How about the acceptance ratio?

Sure I will attach the acc results later. Now we support MTP=1/2/3, but the acceptance rate is low (about 20% for first draft token and 0 for other tokens) and I'm working on it.

File: atom/plugin/vllm/model_wrapper.py [hunk 1]
```diff
@@ -97,7 +137,12 @@ def __init__(self, *, vllm_config: VllmConfig, prefix: str = ""):
 
         _prepare_env(atom_config=self.atom_config)
 
-        model_arch = vllm_config.model_config.architectures[0]
+        main_model_arch = vllm_config.model_config.architectures[0]
+        model_arch = _select_model_arch(vllm_config)
+        self.model_arch = model_arch
+        # if self.forced_model_arch is not None:
+        #     model_arch = self.forced_model_arch
+        #     logger.info(f"Using forced model arch: {model_arch} for vLLM plugin mode")
```
Comment (zejunchen-zejun) [inline atom/plugin/vllm/model_wrapper.py]: should be removed?

File: atom/config.py [hunk 2]
```diff
@@ -932,6 +932,30 @@ def set_current_atom_config(atom_config: Config):
     _current_atom_config = atom_config
 
 
+def _get_current_atom_config_from_vllm_forward_context() -> Optional[Config]:
+    # In vLLM plugin mode (especially speculative decode), main/draft models
+    # can coexist in one process. Resolve per-forward config first to avoid
+    # reading a stale global singleton.
+    if not is_vllm():
+        return None
```
Comment (zejunchen-zejun) [inline atom/config.py]: here should have an assertion to avoid non-vllm backend calling this method
it should only be called by atom-vllm backend

File: atom/config.py [hunk 3]
```diff
@@ -932,6 +932,30 @@ def set_current_atom_config(atom_config: Config):
...
+    if not is_forward_context_available():
+        return None
+    try:
+        return get_vllm_forward_context().additional_kwargs.get("atom_config")
+    except Exception:
+        return None
+
+
 def get_current_atom_config() -> Config:
+    forward_atom_config = _get_current_atom_config_from_vllm_forward_context()
```
Comment (zejunchen-zejun) [inline atom/config.py]: Here maybe a little bit risky. If the `forward_atom_config` is None, and there is no assertion, it will silent fallback to the global singleton `_current_atom_config`. Can we add some log here? Or make it more safe
In ideal situation, the lifecycle and ownership `forward_atom_config` is belong to the model itself and the main model will get its atom config, draft model will get its config. While if draft model cannot get its config, it will fallback to the `_current_atom_config`, which not be correct

File: atom/plugin/vllm/model_wrapper.py [hunk 4]
```diff
@@ -197,6 +303,12 @@ def _register_indexer_caches_with_vllm(self):
                     f"static_forward_context, skipping"
                 )
 
+    def _maybe_adjust_draft_positions(self, positions: torch.Tensor) -> torch.Tensor:
+        position_offset = getattr(self.model, "vllm_draft_position_offset", 0)
+        if position_offset == 0:
+            return positions
+        return positions + position_offset
```
Comment (zejunchen-zejun) [inline atom/plugin/vllm/model_wrapper.py]: can we leave some comments here for the position offset

**Review by ganyi1996ppo — DISMISSED** [review]

File: atom/config.py (hunk 3 above)
Comment (whx-sjtu) [reply atom/config.py]: Do you mean that we should always obtain `forward_atom_config` from `vllm_forward_context`? Will there be scenarios that we need to return the default global `_current_atom_config`?

File: atom/config.py (hunk 2 above)
Comment (whx-sjtu) [reply atom/config.py]: done

File: atom/plugin/vllm/model_wrapper.py (hunk 1 above)
Comment (whx-sjtu) [reply atom/plugin/vllm/model_wrapper.py]: done

File: atom/plugin/vllm/model_wrapper.py (hunk 4 above)
Comment (whx-sjtu) [reply atom/plugin/vllm/model_wrapper.py]: not needed anymore. removed.

File: atom/config.py (hunk 3 above)
Comment (zejunchen-zejun) [reply atom/config.py]: Let us add some comments here to warn a case: there is no atom config in forward context, so the default global config will be provided. With this warning, we can mitigate the coherent issue for local value and its twins global value.

Comment (whx-sjtu) [reply atom/config.py]: Can't add warning log here because this case will be triggered plenty of times while loading model and processing weights. There will be thousands of warning messages logged.

File: atom/plugin/attention.py [hunk 5]
```diff
@@ -2146,16 +2149,27 @@ def init_method_under_plugin_mode(
...
-        )
-        self.reorder_batch_threshold += self.num_speculative_tokens
+        # Determine if this builder is for draft model layers (MTP).
+        # Draft model layers have layer indices >= num_hidden_layers.
+        # The draft model itself does not do speculative decoding, so
+        # num_speculative_tokens should be 0 for its builders.
+        _is_draft_layer = False
+        if layer_names:
+            if len(layer_names) < config.model_config.hf_config.num_hidden_layers:
+                _is_draft_layer = True
```
Comment (zejunchen-zejun) [inline atom/plugin/attention.py]: Here seems a little bit risky to identify a layer is MTP layer or not by comparing the layer numbers. Is it possible to compare the layer index because most time the MTP layer is the last layer of the entire model

**Review by ganyi1996ppo — DISMISSED** [review]

File: atom/plugin/attention.py (hunk 5 above)
Comment (whx-sjtu) [reply atom/plugin/attention.py]: Sure I will try to find a new way here.

File: atom/config.py [hunk 6]
```diff
@@ -1040,6 +1040,31 @@ def set_current_atom_config(atom_config: Config):
     _current_atom_config = atom_config
 
 
+def _get_current_atom_config_from_vllm_forward_context() -> Optional[Config]:
```
Comment (ZhangLirong-amd) [inline atom/config.py:1078]: Can non-VLLM mode return directly without affecting Atom itself?

File: atom/plugin/attention.py [hunk 7]
```diff
@@ -2297,12 +2313,32 @@ def unified_attention_with_output_base_for_plugin_mode(
     use_mla: bool,
     qkv: torch.Tensor,
 ) -> torch.Tensor:
-    atom_config = get_current_atom_config()
+    current_atom_config = None
+    try:
```
Comment (ZhangLirong-amd) [inline atom/plugin/attention.py]: It looks redundant with the updated get_current_atom_config?

File: atom/config.py (hunk 6 above)
Comment (whx-sjtu) [reply atom/config.py:1078]: Sure, thanks for the finding.

File: atom/plugin/attention.py (hunk 7 above)
Comment (whx-sjtu) [reply atom/plugin/attention.py]: Yeah, removed.

File: atom/models/deepseek_mtp.py [hunk 8]
```diff
@@ -88,10 +103,12 @@ def forward(
         spec_step_index: int = 0,
     ) -> torch.Tensor:
         assert inputs_embeds is not None
-        # masked_inputs_embeds = torch.where(
-        #     positions.unsqueeze(-1) == 0, 0, inputs_embeds
-        # )
-        masked_inputs_embeds = inputs_embeds
+        if is_vllm():
```
Comment (zejunchen-zejun) [inline atom/models/deepseek_mtp.py]: Reviewer recommend to move the position code change to atom-vllm model wrapper

**Review by zejunchen-zejun — APPROVED** [review]

**Review by ZhangLirong-amd — APPROVED** [review]

**Review by valarLip — APPROVED** [review]
